### PR TITLE
fix(#285): add composeWithStateSync to resolve issues with enhancer order

### PR DESCRIPTION
**Pull request by Superjo149**
As discussed in #285. Added a new `composeWithStateSync` function which mimics the composer function and should replace that function when using more than one enhancer. When using one enhancer, the other API is still the same.

To accommodate this, I have also added 2 more things:
- `preventActionReplay` was added to the options because in `extensionCompose()` we re-use the `stateSyncEnhancer`, but we do not want the actions to start forwarding there because then it will not pick up the ones dispatched by the middleware. We rather want them to start forwarding in the last part of the chain, hence `forwardActionEnhancer`. This is also possible with denyList, but I thought this was cleaner. WDYT?
- To simplify the compose function, I have opted to check whether in renderer or main process, inside the lib. With that, I have also added a helper function `stateSyncEnhancer` which can wrap around the process specific enhancer to do the same thing.

Will see if I can write a test for it tomorrow. Do you want me to go ahead and update the README already as well @matmalkowski ?

### CODE REVIEW

**Comment by matmalkowski** [comment]
Thanks for the contribution!

We will try to take a look at the change in next few days, you could also add some info into the README file, like you mentioned. I've also started a docs page for the project (`/docs` folder) - you might as well add a paragraph there or just add a placeholder so we can cover it later.

@Slapbox if you have some time, you might also take a look at this branch & try it with your issue #294 and `redux-saga`

**Comment by Slapbox** [comment]
I'd really like to give this a try! 

Unfortunately my TypeScript noobishness may be getting in the way.

I cloned the repo, checked out the `alpha` branch, and ran `yarn && yarn build`, but I get an error like rollup has no idea what TypeScript is. Am I overlooking something obvious?

image

**Comment by Slapbox** [comment]
Friggin' Windows... Builds fine on Linux, will update soon.

**Comment by Slapbox** [comment]
@Superjo149 there's no code changes needed on our end to take advantage of these changes, right?

Unfortunately it doesn't seem to remedy the issue with Redux Saga. 

I took a look at how to setup RxJS and it looks very similar to setting up sagas. Can you share the part of your store configuration where you're applying your enhancers and middlewares so I can compare it with our setup?

**Comment by Superjo149** [comment]
> @Superjo149 there's no code changes needed on our end to take advantage of these changes, right?
> 
> Unfortunately it doesn't seem to remedy the issue with Redux Saga.
> 
> I took a look at how to setup RxJS and it looks very similar to setting up sagas. Can you share the part of your store configuration where you're applying your enhancers and middlewares so I can compare it with our setup?
> 
> I also tried importing `composeWithStateSync` and replacing our `compose` with that, but that leads to `Error: Attempted to register a second handler for 'electron-redux.INIT_STATE_ASYNC'`

@Slapbox There is some kind of code change required, yes. I will properly document this tomorrow along with the tests. But this is the jist.

### Current approach:

```typescript
const middleware = applyMiddleware(countMiddleware)
const enhancer = compose(middleware, mainStateSyncEnhancer())
const store = createStore(reducer, defaultState, enhancer)
```
### Approach with multiple enhancers (Saga middleware,... etc) using composeWithStateSync:
> Note that when using `composeWithStateSync`, neither `mainStateSyncEnhancer` or `rendererStateSyncEnhancer` are required as this will be handled by the compose function. This in line with how `redux-devtools-extension` works.
 
```typescript
const middleware = applyMiddleware(countMiddleware)
const enhancer = composeWithStateSync(middleware)
const store = createStore(reducer, defaultState, enhancer)
```

**Comment by Slapbox** [comment]
@Superjo149 thanks so much for your help, that makes perfect sense now and I can confirm that everything seems to work properly!

Not only a fix, but a further streamlining of an already great API; wow, great work!

**Comment by Slapbox** [comment]
Oh, you can just wrap compose functions around compose functions. Never knew that, so that's the solution for Redux Devtools in case anyone is lurking this thread.

File: yarn.lock [hunk 1]
```diff
@@ -4503,6 +4503,11 @@ electron-devtools-installer@^2.2.4:
     rimraf "^2.5.2"
     semver "^5.3.0"
 
+electron-process-type@^4.2.0:
```
Comment (matmalkowski) [inline yarn.lock:4506]: I'm not entirely sure if this is the way to go - I like the convenience though of having single function taking care of it - but library itself doesn't seem very popular / maintained (with only single contributor) & its for some reason heavier than this entire package :D Maybe we can extract that logic? 🤔

File: src/utils/actions.ts [hunk 2]
```diff
@@ -9,7 +9,7 @@ export type ActionMeta = {
  * stopForwarding allows you to give it an action, and it will return an
  * equivalent action that will only play in the current process
  */
-export const stopForwarding = (action: FluxStandardAction<ActionMeta>) => ({
+export const stopForwarding = (action: FluxStandardAction<ActionMeta>): any => ({
```
Comment (matmalkowski) [inline src/utils/actions.ts:12]: Why explicit `any`?

File: src/composeWithStateSync.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/* eslint-disable @typescript-eslint/ban-types */
+
+import { StoreEnhancer, StoreEnhancerStoreCreator } from 'redux'
+import { forwardAction } from './forwardAction'
+import { MainStateSyncEnhancerOptions } from './options/MainStateSyncEnhancerOptions'
+import { RendererStateSyncEnhancerOptions } from './options/RendererStateSyncEnhancerOptions'
+import { stateSyncEnhancer } from './stateSyncEnhancer'
+
+export type StateSyncOptions = MainStateSyncEnhancerOptions | RendererStateSyncEnhancerOptions
```
Comment (matmalkowski) [inline src/composeWithStateSync.ts:9]: This looks like a circular type dependency to me => `MainStateSyncEnhancerOptions` extends the `StateSyncOptions` in the other place, yet here it is an union?

**Review by matmalkowski — CHANGES_REQUESTED** [review]
This looks exiting! I believe this is the right direction, and will get us closer to releasing stable v2 version, thanks a lot for all the hard work you put into figuring out how to implement this 👍🏻

File: yarn.lock (hunk 1 above)
Comment (Slapbox) [reply yarn.lock:4506]: This was my only qualm.

The only code we need to determine `renderer` versus `main` is:

```js
export const IS_MAIN_PROCESS = process.type === 'browser';
export const IS_RENDERER_PROCESS = process.type === 'renderer';
```

It mentions different results for sandboxing and other situations though, so I'm not sure my suggestion is as complete - but I think it's complete enough.

Comment (Superjo149) [reply yarn.lock:4506]: I was hesitant to use this package, but I couldn't say with 100% certainty that `process.type` would always properly be set. I think we can just use process.type and remove the package and see from there.

File: src/utils/actions.ts (hunk 2 above)
Comment (Superjo149) [reply src/utils/actions.ts:12]: I was having a hard time trying to make the store typings comply, the was the only one which I could't find a solution for.

File: src/composeWithStateSync.ts (hunk 3 above)
Comment (Superjo149) [reply src/composeWithStateSync.ts:9]: Good find, this was "legacy" and can be removed.

File: src/utils/actions.ts (hunk 2 above)
Comment (Slapbox) [reply src/utils/actions.ts:12]: Is this the last obstacle to merging this PR?

**Review by matmalkowski — APPROVED** [review]

**Comment by Superjo149** [comment]
Glad I could contribute to this 🎉

**Pull request merged**